File: Shantanu_Mamgain/Sliding-Window-Leetcode/Minimum Window Substring.java
class Solution {
    public String minWindow(String s, String t) {
        int start = 0;
        int end = 0;
        int minLen = s.length()+1;
        int si = -1;
        int ei = -1;
        int count = 0;

        char s1[] = new char [128];
        char t1[] = new char [128];

        for(int i = 0; i < t.length(); i++){
            t1[t.charAt(i)] += 1;
            if(t1[t.charAt(i)] == 1){
                count ++;
            }
        }
        while(end < s.length()){
            char c=s.charAt(end);
            s1[c]++;
            if(s1[c] == t1[c]){
                count --;
            }
            while(count == 0){
                if(minLen > end - start + 1){
                    minLen = end-start+1;
                    si = start;
                    ei = end;
                }
                char ch=s.charAt(start);
                if(s1[ch] == t1[ch]){
                    count++;
                }
                s1[ch] --;
                start ++;
            }
            end ++;
        }

        if(si==-1){
            return "";
        }
        else{
            return s.substring(si,si+minLen);
        }
    }
}

File: NikhilKumarGupta/Slidingwindow/Minimum Window Substring.cpp
class Solution {
    public:
        string minWindow(string s, string t) {
            vector<int>tfreq(128,0);
            vector<int>window(128,0);
            int left=0;
            int right=0;
            int start=0;
            int match=0;
            int mini=INT_MAX;
    
            for(int i=0;i<t.size();i++){
                tfreq[t[i]]++;
            }
            for(int i=0;i<s.size();i++){
                window[s[right]]++;
    
                if(tfreq[s[right]]>0&&window[s[right]]<=tfreq[s[right]]){
                    match++;
                }
                while(match==t.size()){
                    if(mini>right-left+1){
                        mini=right-left+1;
                        start=left;
                    }
                    if(tfreq[s[left]]>0&&window[s[left]]==tfreq[s[left]]){
                    match--;
                }
                 window[s[left]]--;
                left++;
                }
                right++;
             
               
            }
            return mini==INT_MAX ?"":s.substr(start,mini);
            
        }
    };